kontrol/crate.go

```diff
@@ -0,0 +1,205 @@
+package kontrol
+
+import (
+	"database/sql"
+	"fmt"
+
+	_ "github.com/herenow/go-crate"
+
+	"github.com/koding/kite"
+	kontrolprotocol "github.com/koding/kite/kontrol/protocol"
+	"github.com/koding/kite/protocol"
+	"github.com/koding/multiconfig"
+)
+
+const (
+	CratePrefix = "kontrol_crate"
+)
+
+// Crate holds Crate database related configuration
+type CrateConfig struct {
+	Host  string `default:"127.0.0.1"`
+	Port  int    `default:"4200"`
+	Table string `default:"kontrol"`
+}
+
+type Crate struct {
+	DB    *sql.DB
+	Log   kite.Logger
+	Table string
+}
+
+func NewCrate(conf *CrateConfig, log kite.Logger) *Crate {
+	if conf == nil {
+		conf = new(CrateConfig)
+
+		envLoader := &multiconfig.EnvironmentLoader{Prefix: CratePrefix}
+		configLoader := multiconfig.MultiLoader(
+			&multiconfig.TagLoader{}, envLoader,
+		)
+
+		if err := configLoader.Load(conf); err != nil {
+			log.Error("Valid environment variables are: ")
+			envLoader.PrintEnvs(conf)
+			log.Fatal("%v", err)
+		}
+
+		err := multiconfig.MultiValidator(&multiconfig.RequiredValidator{}).Validate(conf)
+		if err != nil {
+			log.Error("Valid environment variables are: ")
+			envLoader.PrintEnvs(conf)
+			log.Fatal("%v", err)
+		}
+	}
+
+	connString := fmt.Sprintf("http://%s:%d/", conf.Host, conf.Port)
+
+	db, err := sql.Open("crate", connString)
+	if err != nil {
+		log.Fatal("%v", err)
+	}
+
+	c := &Crate{
+		DB:    db,
+		Log:   log,
+		Table: conf.Table,
+	}
+
+	// TODO
+	// cleanInterval := 120 * time.Second // clean every 120 second
+	// go p.RunCleaner(cleanInterval, KeyTTL)
+
+	return c
+}
+
+// Get retrieves the Kites with the given query
+func (c *Crate) Get(query *protocol.KontrolQuery) (Kites, error) {
+	return nil, fmt.Errorf("Not Impmentented")
+	// // We will make a get request to etcd store with this key. So get a "etcd"
```

Comment:
This should kites from crate directly, design-wise storages should be self-contained without dependencies.

If we wanted to read part from one storage and part from another, we create another storage, that would compose those storages. Like io.Writer and io.Readers are composed.